modules/drivers/mongo/src/metabase/driver/mongo.clj

```diff
@@ -553,6 +554,144 @@
+  {:arglists '([type])}
+  identity)
+
+(defmethod type->database-type :type/TextLike [_] "string")
+(defmethod type->database-type :type/Text [_] "string")
+(defmethod type->database-type :type/Number [_] "long")
+(defmethod type->database-type :type/Integer [_] "int")
+(defmethod type->database-type :type/BigInteger [_] "long")
+(defmethod type->database-type :type/Float [_] "double")
+(defmethod type->database-type :type/Decimal [_] "decimal")
+(defmethod type->database-type :type/Boolean [_] "bool")
+(defmethod type->database-type :type/Date [_] "date")
+(defmethod type->database-type :type/DateTime [_] "date")
+(defmethod type->database-type :type/DateTimeWithTZ [_] "date")
+(defmethod type->database-type :type/Time [_] "date")
+(defmethod type->database-type :type/TimeWithTZ [_] "date")
+(defmethod type->database-type :type/Instant [_] "date")
+(defmethod type->database-type :type/UUID [_] "uuid")
+(defmethod type->database-type :type/JSON [_] "object")
+(defmethod type->database-type :type/SerializedJSON [_] "string")
+(defmethod type->database-type :type/Array [_] "array")
+(defmethod type->database-type :type/Dictionary [_] "object")
+(defmethod type->database-type :type/MongoBSONID [_] "objectId")
+(defmethod type->database-type :type/MongoBinData [_] "binData")
+(defmethod type->database-type :type/IPAddress [_] "string")
+(defmethod type->database-type :default [_] "object")
+
+(defmethod driver/type->database-type :mongo
+  [_driver base-type]
+  (type->database-type base-type))
+
+(defn- convert-value-for-insertion
+  [base-type value]
+  (condp #(isa? %2 %1) base-type
+    :type/JSON
+    (json/decode value)
+
+    :type/Dictionary
+    (json/decode value)
+
+    :type/Array
+    (json/decode value)
+
+    :type/Integer
+    (parse-long value)
+
+    :type/BigInteger
+    (bigint value)
+
+    :type/Float
+    (parse-double value)
+
+    :type/Decimal
+    (bigdec value)
+
+    :type/Number
+    (bigint value)
+
+    :type/Boolean
+    (parse-boolean value)
+
+    :type/Date
+    (u.date/parse value)
+
+    :type/DateTime
+    (u.date/parse value)
+
+    :type/DateTimeWithTZ
+    (u.date/parse value)
+
+    :type/Time
+    (u.date/parse value)
+
+    :type/TimeWithTZ
+    (u.date/parse value)
+
+    :type/Instant
+    (u.date/parse value)
+
+    :type/UUID
+    (parse-uuid value)
+
+    value))
+
+(defmethod driver/string->val :mongo
+  [_driver column-def string-val]
+  (convert-value-for-insertion (:type column-def) string-val))
+
+(defmethod driver/insert-from-source! [:mongo :rows]
+  [_driver db-id {table-name :name :keys [columns]} {:keys [data]}]
+  (let [col-names (mapv :name columns)]
+    (mongo.connection/with-mongo-database [^MongoDatabase db db-id]
+      (let [collection (mongo.util/collection db (name table-name))
+            documents (map #(into {} (map vector col-names %))
+                           data)]
+        (if (> (bounded-count 2 documents) 1)
+          (mongo.util/insert-many collection documents)
```

Comment:
deep in the bowels of mongo pretty sure [this](https://github.com/mongodb/mongo-java-driver/blob/347a6e6c5a4b50d773d0982609e1a1739d714553/driver-core/src/main/com/mongodb/internal/operation/Operations.java#L431) forces the seq.

Reply:
insert-many (I know this is not new) could make it clear it is best used on small inputs.

Here I think I guess could either comment 'not enough time / doesnt matter right now as python transforms have to be small anyway' or do some kind of loop (likely not without its own issues).

🤷

Reply:
are you suggestion we loop over partitions instead?

Reply:
> are you suggestion we loop over partitions instead?

I would probably leave the comment for now as we already realize things in memory

Reply:
left TODO